### DISCUSSION TITLE
TokenSet with expires_in (instead of expires_at)

### DISCUSSION THREAD

**Discussion by GianlucaDvorak** [panva/openid-client#634] [Q&A]
Hello together,

We facing an issue with the TokenSetParameters.
At the moment we receive as a response a totally fine token with the parameters:

    "access_token": 
    "token_type":
    "id_token": 
    "refresh_token":
    "expires_at": 
    "scope": "openid",
    "jti": 

But now we would like to get the expires_in parameter in the tokenset.
Is there a way to customize the tokenset?

Thank you
Gian

**Comment by panva** [comment]
Hi @GianlucaDvorak,

I am unsure about what you're asking for. The `expires_in` is by nature a [calculated getter](https://github.com/panva/node-openid-client/blob/v5.x/lib/token_set.js#L18-L20).

**Reply by GianlucaDvorak** [reply]
Hello @panva 

When we using the client.callback(redirectUri, parameters[, checks[, extras]]) it returns the TokenSet.
The returned TokenSet looks like the structure above (without the attribute expires_in). I'm also not able to modify the TokenSet with adding the attribute "expires_in". So my question is, if I can somehow modify the returned TokenSet from the client.callback (exchange the attributes expires_at and expires_in).

**Reply by panva** [reply]
It's not a serializable attribute, it's a getter and a setter pair that's based on the expires_at which is again calculated at the time the token endpoint response was received IF it contained the OPTIONAL expires_in property. So in case you don't have neither expires_at/expires_in, it's because the issuer didn't provide you one.

**Reply by GianlucaDvorak** [reply]
Thanks for you explanation.
I get a expires_in from my idp service.

But I can probably show you my issue with a test function from your code.
In the node-openid-client/test/tokenset/tokenset.test.js you coded a test function on line 68 to 71:

    ts = new TokenSet({ expires_in: 'foo' });
    ts.expires_in = 200;
    expect(ts.expires_in).to.be.a('number');
    expect(ts.expired()).to.eql(false);

When I use this function in my code, the test will fail. So the ts does not include an attribute expires_in instead it has an attribute expires_at (according to the setter function of TokenSet). So the attribute expires_in will never be possible to add to the TokenSet.

**Reply by panva** [reply]
> When I use this function in my code, the test will fail.

I dont know what to say, when the tests run and when i try the same in my REPL it works as expected. Sounds like an issue on your end.

**Reply by GianlucaDvorak** [reply]
You're right. But the ts.expires_in is just a getter method, which is calculated during the runtime and it is no fixed attribute in the tokenset. We need expires_in as an attribute in the tokenset itself. The reason behind that is, that the service provider needs the attribute expries_in in the tokenset, which he receives from this service. At the moment our service provider only receives the expries_at, because the expires_in is visible only inside this oauth service.
